### PR TITLE
TST: Add a fixture to close all plots

### LINKED ISSUE #4227: TST: closing plots, pytest teardown
we have a large number of plots that are not properly closed
see https://github.com/statsmodels/statsmodels/pull/4226#issuecomment-360821013

in gofplots:

Mixin class has a `teardown` method, however that is not a recognized pytest name. pytest uses either a class method `teardown_class` or a regular method `teardown_method` which is called after every test method. test_gofplots doesn't have any `close` in the test methods directly.

The only other teardown in graphics.tests is in regression_plots which uses `teardown_module` in addition to the close in individual test methods. So that looks good.

statespace:
AFAICS from a file search, all plots have a plt.close(fig) after the call to the plot method.

### PR DESCRIPTION
**Pull request by bashtage**
Add a fixture that can be used by any plotting function to close
all plots when the function exits, whether it succeeds or fails.
Switch plotting test to use the fixture.

closes #4227

### CODE REVIEW

File: conftest.py [hunk 1]
```diff
@@ -19,3 +19,24 @@ def pytest_runtest_setup(item):
...
+
+    def test_some_plot(close_figures):
+        <test code>
+    """
+    yield None
+    try:
+        from matplotlib.pyplot import close
+        close('all')
+    except ImportError:
+        pass
```
Comment (jbrockmendel) [inline conftest.py]: Newline

**Comment by josef-pkt** [comment]
I need to run this locally before merging. 
I got in the past a large number of segfaults in matplotlib especially if figures where not closed right away. In those cases a `close('all')` at the end of the test function was not enough when several plots were kept open in a function or method.

**Comment by josef-pkt** [comment]
ok, segfault
Before checking out this PR branch, the graphics tests passed and the entire test suite ran without segfault.

```
>pytest  "...\statsmodels_py34_pr\statsmodels\graphics\tests"
============================= test session starts =============================
platform win32 -- Python 3.4.4, pytest-3.0.3, py-1.4.31, pluggy-0.4.0
rootdir: ...\statsmodels_py34_pr, inifile:
collected 128 items

..\eclipse_ws\statsmodels\statsmodels_py34_pr\statsmodels\graphics\tests\test_ag
reement.py .

<segfault>
```

@bashtage  I have had segfaults with matplotlib for several years, and the close as they are in master where added to avoid segfaults for many different cases.
In some cases, the close were not enough to avoid the segfaults, and then I changed the matplotlib backend in the test module.

My guess is that there is a problem with the TK or QT backend, and plots don't get properly closed.
Maybe switching the matplotlib backend for all tests would help.

Essentially, as I see it, the current version works, and all changes to it need to make sure all plots are properly closed and if possible as soon as they are not needed anymore.

**Comment by bashtage** [comment]
I’m surprised.  I manually verified the number of figures was always 0 the end of every test.

I suspect that the path matters. To get the conftest.py to work as expect I think you need to be running the test from the root of the repo.  Your command makes it look like you were somewhere else.

Might need to relocate conftest to statsmodels\statsmodels\conftest which is what pandas does.



________________________________

**Comment by josef-pkt** [comment]
Changing the backend looks good so far, both the `graphics` and `tsa` tests finished without segfaulting and both were segfaulting before the update.

**Comment by josef-pkt** [comment]
I ran the entire test suite again and there are no segfaults.

**Comment by josef-pkt** [comment]
@bashtage thank you

**Pull request merged**